File: src/main/java/aurocosh/divinefavor/common/entity/rope/EntityRopeTeleportingNode.java
package aurocosh.divinefavor.common.entity.rope;

import aurocosh.divinefavor.common.entity.rope.base.EntityRopeNodeBase;
import aurocosh.divinefavor.common.item.common.ModItems;
import aurocosh.divinefavor.common.util.SlotData;
import aurocosh.divinefavor.common.util.UtilCoordinates;
import aurocosh.divinefavor.common.util.UtilEntity;
import aurocosh.divinefavor.common.util.UtilPlayer;
import net.minecraft.entity.Entity;
import net.minecraft.entity.EntityLivingBase;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.item.ItemStack;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.World;

public class EntityRopeTeleportingNode extends EntityRopeNodeBase {
    public EntityRopeTeleportingNode(World world) {
        super(world);
    }

    @Override
    protected void registerPickUp(EntityPlayer player) {
        UtilPlayer.addStackToInventoryOrDrop(player, new ItemStack(ModItems.INSTANCE.getRope_teleporting(), 1));
    }

    @Override
    protected boolean canDropNewNode(EntityPlayer player) {
        SlotData slotData = UtilPlayer.findStackInInventory(player, element -> !element.isEmpty() && element.getItem() == ModItems.INSTANCE.getRope_teleporting());
        if (slotData.getSlotIndex() == -1)
            return false;
        slotData.getStack().shrink(1);
        player.inventory.setInventorySlotContents(slotData.getSlotIndex(), slotData.getStack().getCount() > 0 ? slotData.getStack() : ItemStack.EMPTY);
        return true;
    }

    @Override
    protected EntityRopeNodeBase makeNewNode(World world) {
        return new EntityRopeTeleportingNode(world);
    }

    @Override
    protected Class<? extends EntityRopeNodeBase> getEntityClass() {
        return EntityRopeTeleportingNode.class;
    }

    @Override
    public boolean canBeCollidedWith() {
        return true;
    }

    @Override
    public boolean canBePushed() {
        return true;
    }

    @Override
    public void applyEntityCollision(Entity entityIn) {
        if (!(entityIn instanceof EntityLivingBase))
            return;
        EntityLivingBase livingBase = (EntityLivingBase) entityIn;
        EntityRopeNodeBase lastConnectedNode = getLastConnectedNode(this);
        if (lastConnectedNode == this)
            return;
        BlockPos placeToStand = UtilCoordinates.INSTANCE.findPlaceToStand(lastConnectedNode.getPosition(), world, 8);
        UtilEntity.INSTANCE.teleport(livingBase, placeToStand);
    }
}

File: src/main/kotlin/aurocosh/divinefavor/common/entity/rope/EntityRopeTeleportingNode.kt
package aurocosh.divinefavor.common.entity.rope

import aurocosh.divinefavor.common.entity.rope.base.EntityRopeNodeBase
import aurocosh.divinefavor.common.item.common.ModItems
import aurocosh.divinefavor.common.util.UtilCoordinates
import aurocosh.divinefavor.common.util.UtilEntity
import aurocosh.divinefavor.common.util.UtilPlayer
import net.minecraft.entity.Entity
import net.minecraft.entity.EntityLivingBase
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.item.ItemStack
import net.minecraft.world.World

class EntityRopeTeleportingNode(world: World) : EntityRopeNodeBase(world) {

    override val entityClass: Class<out EntityRopeNodeBase>
        get() = EntityRopeTeleportingNode::class.java

    override fun registerPickUp(player: EntityPlayer) {
        UtilPlayer.addStackToInventoryOrDrop(player, ItemStack(ModItems.rope_teleporting, 1))
    }

    override fun canDropNewNode(player: EntityPlayer): Boolean {
        val (slotIndex, stack) = UtilPlayer.findStackInInventory(player) { element -> !element.isEmpty && element.item === ModItems.rope_teleporting }
        if (slotIndex == -1)
            return false
        stack.shrink(1)
        player.inventory.setInventorySlotContents(slotIndex, if (stack.count > 0) stack else ItemStack.EMPTY)
        return true
    }

    override fun makeNewNode(world: World): EntityRopeNodeBase {
        return EntityRopeTeleportingNode(world)
    }

    override fun canBeCollidedWith(): Boolean {
        return true
    }

    override fun canBePushed(): Boolean {
        return true
    }

    override fun applyEntityCollision(entityIn: Entity) {
        if (entityIn !is EntityLivingBase)
            return
        val lastConnectedNode = getLastConnectedNode(this)
        if (lastConnectedNode === this)
            return
        val placeToStand = UtilCoordinates.findPlaceToStand(lastConnectedNode.position, world, 8)
        UtilEntity.teleport(entityIn, placeToStand!!)
    }
}